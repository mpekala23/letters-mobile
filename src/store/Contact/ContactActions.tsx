import {
  Contact,
  SET_ADDING,
  SET_EXISTING,
  CLEAR_CONTACTS,
  ContactActionTypes,
} from "./ContactTypes";

export function setAdding(contact: Contact): ContactActionTypes {
  return {
    type: SET_ADDING,
    payload: contact,
  };
}

export function setExisting(contacts: Contact[]): ContactActionTypes {
  return {
    type: SET_EXISTING,
    payload: contacts,
  };
}

export function clearContacts() {
  return {
    type: CLEAR_CONTACTS,
    payload: null,
  };
}
