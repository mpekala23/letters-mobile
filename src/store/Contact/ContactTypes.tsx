import { Facility } from "types";

export const SET_ADDING = "contact/set_adding";
export const SET_EXISTING = "contact/set_existing";
export const CLEAR_CONTACTS = "contact/clear_contacts";

export interface Contact {
  id: number;
  state: string;
  firstName: string;
  lastName: string;
  inmateNumber: string;
  relationship: string;
  facility: Facility | null;
}

export interface ContactState {
  adding: Contact;
  existing: Contact[];
}

interface SetAddingAction {
  type: typeof SET_ADDING;
  payload: Contact;
}

interface SetExistingAction {
  type: typeof SET_EXISTING;
  payload: Contact[];
}

interface ClearContactsAction {
  type: typeof CLEAR_CONTACTS;
  payload: null;
}

export type ContactActionTypes =
  | SetAddingAction
  | SetExistingAction
  | ClearContactsAction;
