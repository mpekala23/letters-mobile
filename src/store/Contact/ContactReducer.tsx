import { ContactState, ContactActionTypes, SET_ADDING, SET_EXISTING } from './ContactTypes';

const initialState: ContactState = {
  adding: {
    state: '',
    first_name: '',
    last_name: '',
    inmate_number: '',
    relationship: '',
    facility: null,
  },
  existing: [],
};

export default function ContactReducer(
  state = initialState,
  action: ContactActionTypes
): ContactState {
  const currentState = { ...state };
  switch (action.type) {
    case SET_ADDING:
      currentState.adding = action.payload;
      return currentState;
    case SET_EXISTING:
      currentState.existing = action.payload;
      return currentState;
    default:
      return state;
  }
}
