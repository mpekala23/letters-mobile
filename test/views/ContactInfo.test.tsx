import * as React from "react";
import { ContactInfoScreen } from "@views";
import { render, toJSON, fireEvent } from "@testing-library/react-native";
import { SET_ADDING } from "@store/Contact/ContactTypes";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);

const setup = (contactOverrides = {}) => {
  const navigation = { navigate: jest.fn(), addListener: jest.fn() };
  const contact = Object.assign(
    {
      state: "",
      firstName: "",
      lastName: "",
      inmateNumber: "",
      relationship: "",
      facility: null,
    },
    contactOverrides
  );
  const initialState = {
    adding: contact,
    existing: [],
  };
  const store = mockStore({
    contact: initialState,
  });

  const StoreProvider = ({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  };

  return {
    navigation,
    store,
    ...render(
      <ContactInfoScreen navigation={navigation} contactState={initialState} />,
      {
        wrapper: StoreProvider,
      }
    ),
  };
};

describe("Contact Info Screen", () => {
  it("should match snapshot", () => {
    const { container } = setup(<ContactInfoScreen />);
    const tree = toJSON(container);
    expect(tree).toMatchSnapshot();
  });

  it("should have next button be disabled until all fields are valid", () => {
    const { navigation, getByPlaceholderText, getByText } = setup();
    const nextButton = getByText("Next");
    fireEvent.press(nextButton);
    expect(
      getByText("Next").parentNode.props.style[1].backgroundColor
    ).toBeDefined();
    expect(navigation.navigate).toHaveBeenCalledTimes(0);
    fireEvent.changeText(getByPlaceholderText("State"), "Minnesota");
    fireEvent.changeText(getByPlaceholderText("First Name"), "First");
    fireEvent.changeText(getByPlaceholderText("Last Name"), "Last");
    fireEvent.changeText(getByPlaceholderText("Inmate Number"), "2");
    fireEvent.changeText(
      getByPlaceholderText("Relationship to Inmate"),
      "Mother"
    );
    fireEvent.press(nextButton);
    expect(getByText("Next").parentNode.props.style[1]).toEqual({});
    expect(navigation.navigate).toHaveBeenCalledTimes(1);
  });

  it("should dispatch contact info to the redux store when the next button is pressed", () => {
    const { store, getByPlaceholderText, getByText } = setup();
    const nextButton = getByText("Next");
    fireEvent.press(nextButton);
    expect(store.getActions().length).toBe(0);
    fireEvent.changeText(getByPlaceholderText("State"), "Minnesota");
    fireEvent.changeText(getByPlaceholderText("First Name"), "First");
    fireEvent.changeText(getByPlaceholderText("Last Name"), "Last");
    fireEvent.changeText(getByPlaceholderText("Inmate Number"), "2");
    fireEvent.changeText(
      getByPlaceholderText("Relationship to Inmate"),
      "Mother"
    );
    fireEvent.press(nextButton);
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SET_ADDING,
      payload: {
        state: "Minnesota",
        firstName: "First",
        lastName: "Last",
        inmateNumber: "2",
        relationship: "Mother",
        facility: null,
      },
    });
  });

  it("should navigate to the facility directory screen when the next button is pressed", () => {
    const { navigation, getByPlaceholderText, getByText } = setup();
    const nextButton = getByText("Next");
    fireEvent.changeText(getByPlaceholderText("State"), "Minnesota");
    fireEvent.changeText(getByPlaceholderText("First Name"), "First");
    fireEvent.changeText(getByPlaceholderText("Last Name"), "Last");
    fireEvent.changeText(getByPlaceholderText("Inmate Number"), "2");
    fireEvent.changeText(
      getByPlaceholderText("Relationship to Inmate"),
      "Mother"
    );
    fireEvent.press(nextButton);
    expect(navigation.navigate).toHaveBeenCalledWith("FacilityDirectory");
  });

  it("should load initial values for fields from the redux store", () => {
    const { getByPlaceholderText } = setup({
      state: "Minnesota",
      firstName: "First",
      lastName: "Last",
      inmateNumber: "6",
      relationship: "Sister",
    });
    expect(getByPlaceholderText("State").props.value).toBe("Minnesota");
    expect(getByPlaceholderText("First Name").props.value).toBe("First");
    expect(getByPlaceholderText("Last Name").props.value).toBe("Last");
    expect(getByPlaceholderText("Inmate Number").props.value).toBe("6");
    expect(getByPlaceholderText("Relationship to Inmate").props.value).toBe(
      "Sister"
    );
  });
});
