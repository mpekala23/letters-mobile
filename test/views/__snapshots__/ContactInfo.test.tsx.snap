// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Info Screen should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TouchableOpacity
    activeOpacity={1}
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <ScrollView
          keyboardShouldPersistTaps="handled"
          scrollEnabled={false}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "height": 40,
                "width": "100%",
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "white",
                "flex": 1,
                "paddingHorizontal": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "backgroundColor": "#3490dc",
                    "borderBottomRightRadius": 5,
                    "borderTopRightRadius": 5,
                    "height": 10,
                    "marginRight": 10,
                    "width": 20,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "fontFamily": "Poppins-Bold",
                    "fontSize": 30,
                  }
                }
              >
                Add Contact
              </Text>
            </View>
            <TouchableOpacity
              activeOpacity={0.2}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "marginTop": 10,
                }
              }
              testID="clickable"
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins-Regular",
                    },
                    Object {
                      "color": "#3490dc",
                      "textDecorationLine": "underline",
                    },
                  ]
                }
              >
                Need help finding your inmate ID?
              </Text>
            </TouchableOpacity>
            <TouchableOpacity
              activeOpacity={0.2}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "marginTop": 10,
                }
              }
              testID="clickable"
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins-Regular",
                    },
                    Object {
                      "color": "#3490dc",
                      "textDecorationLine": "underline",
                    },
                  ]
                }
              >
                <Text>
                  Tap here to search the
                   
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "Poppins-Bold",
                        },
                        Object {
                          "color": "#282828",
                        },
                      ]
                    }
                  >
                    CT
                  </Text>
                   
                  database
                  .
                </Text>
              </Text>
            </TouchableOpacity>
            <TouchableOpacity
              activeOpacity={0.2}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "marginBottom": 30,
                  "marginTop": 10,
                }
              }
              testID="clickable"
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins-Regular",
                    },
                    Object {
                      "color": "#3490dc",
                      "textDecorationLine": "underline",
                    },
                  ]
                }
              >
                <Text>
                  Tap here to search the
                   
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "Poppins-Bold",
                        },
                        Object {
                          "color": "#282828",
                        },
                      ]
                    }
                  >
                    Federal
                  </Text>
                   
                  database
                  .
                </Text>
              </Text>
            </TouchableOpacity>
            <View
              pointerEvents="auto"
              style={
                Object {
                  "backgroundColor": "#F2F2F2",
                  "borderRadius": 8,
                  "height": 60,
                  "marginBottom": 10,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
              testID="parent"
            >
              <View
                testID="invalid"
              />
              <View
                testID="unfocused"
              />
              <ScrollView
                keyboardShouldPersistTaps="always"
                scrollEnabled={false}
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "width": "100%",
                    },
                    Object {},
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  multiline={false}
                  numberOfLines={1}
                  placeholder="State"
                  rejectResponderTermination={true}
                  secureTextEntry={false}
                  style={
                    Array [
                      Object {
                        "borderColor": "rgba(0,0,0,0)",
                        "borderRadius": 8,
                        "borderWidth": 2,
                        "color": "#9A9A9A",
                        "flex": 1,
                        "fontSize": 16,
                        "padding": 10,
                        "width": "100%",
                      },
                      Object {},
                      Object {},
                      Object {
                        "height": 50,
                      },
                      Object {},
                    ]
                  }
                  underlineColorAndroid="transparent"
                  value=""
                />
                <View />
                <View />
                <View
                  style={
                    Object {
                      "height": 200,
                      "width": "100%",
                      "zIndex": 5,
                    }
                  }
                >
                  <View />
                </View>
              </ScrollView>
            </View>
            <View
              pointerEvents="auto"
              style={
                Object {
                  "backgroundColor": "#F2F2F2",
                  "borderRadius": 8,
                  "height": 50,
                  "marginBottom": 10,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
              testID="parent"
            >
              <View
                testID="invalid"
              />
              <View
                testID="unfocused"
              />
              <ScrollView
                keyboardShouldPersistTaps="always"
                scrollEnabled={false}
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "width": "100%",
                    },
                    Object {},
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  multiline={false}
                  numberOfLines={1}
                  placeholder="First Name"
                  rejectResponderTermination={true}
                  secureTextEntry={false}
                  style={
                    Array [
                      Object {
                        "borderColor": "rgba(0,0,0,0)",
                        "borderRadius": 8,
                        "borderWidth": 2,
                        "color": "#9A9A9A",
                        "flex": 1,
                        "fontSize": 16,
                        "padding": 10,
                        "width": "100%",
                      },
                      Object {},
                      Object {},
                      Object {
                        "height": 50,
                      },
                      Object {},
                    ]
                  }
                  underlineColorAndroid="transparent"
                  value=""
                />
                <View />
                <View />
                <View
                  style={
                    Object {
                      "height": 0,
                      "width": "100%",
                      "zIndex": 5,
                    }
                  }
                >
                  <View />
                </View>
              </ScrollView>
            </View>
            <View
              pointerEvents="auto"
              style={
                Object {
                  "backgroundColor": "#F2F2F2",
                  "borderRadius": 8,
                  "height": 50,
                  "marginBottom": 10,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
              testID="parent"
            >
              <View
                testID="invalid"
              />
              <View
                testID="unfocused"
              />
              <ScrollView
                keyboardShouldPersistTaps="always"
                scrollEnabled={false}
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "width": "100%",
                    },
                    Object {},
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  multiline={false}
                  numberOfLines={1}
                  placeholder="Last Name"
                  rejectResponderTermination={true}
                  secureTextEntry={false}
                  style={
                    Array [
                      Object {
                        "borderColor": "rgba(0,0,0,0)",
                        "borderRadius": 8,
                        "borderWidth": 2,
                        "color": "#9A9A9A",
                        "flex": 1,
                        "fontSize": 16,
                        "padding": 10,
                        "width": "100%",
                      },
                      Object {},
                      Object {},
                      Object {
                        "height": 50,
                      },
                      Object {},
                    ]
                  }
                  underlineColorAndroid="transparent"
                  value=""
                />
                <View />
                <View />
                <View
                  style={
                    Object {
                      "height": 0,
                      "width": "100%",
                      "zIndex": 5,
                    }
                  }
                >
                  <View />
                </View>
              </ScrollView>
            </View>
            <View
              pointerEvents="auto"
              style={
                Object {
                  "backgroundColor": "#F2F2F2",
                  "borderRadius": 8,
                  "height": 50,
                  "marginBottom": 10,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
              testID="parent"
            >
              <View
                testID="invalid"
              />
              <View
                testID="unfocused"
              />
              <ScrollView
                keyboardShouldPersistTaps="always"
                scrollEnabled={false}
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "width": "100%",
                    },
                    Object {},
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  multiline={false}
                  numberOfLines={1}
                  placeholder="Inmate Number"
                  rejectResponderTermination={true}
                  secureTextEntry={false}
                  style={
                    Array [
                      Object {
                        "borderColor": "rgba(0,0,0,0)",
                        "borderRadius": 8,
                        "borderWidth": 2,
                        "color": "#9A9A9A",
                        "flex": 1,
                        "fontSize": 16,
                        "padding": 10,
                        "width": "100%",
                      },
                      Object {},
                      Object {},
                      Object {
                        "height": 50,
                      },
                      Object {},
                    ]
                  }
                  underlineColorAndroid="transparent"
                  value=""
                />
                <View />
                <View />
                <View
                  style={
                    Object {
                      "height": 0,
                      "width": "100%",
                      "zIndex": 5,
                    }
                  }
                >
                  <View />
                </View>
              </ScrollView>
            </View>
            <View
              pointerEvents="auto"
              style={
                Object {
                  "backgroundColor": "#F2F2F2",
                  "borderRadius": 8,
                  "height": 60,
                  "marginBottom": 0,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
              testID="parent"
            >
              <View
                testID="invalid"
              />
              <View
                testID="unfocused"
              />
              <ScrollView
                keyboardShouldPersistTaps="always"
                scrollEnabled={false}
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "width": "100%",
                    },
                    Object {},
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  multiline={false}
                  numberOfLines={1}
                  placeholder="Relationship to Inmate"
                  rejectResponderTermination={true}
                  secureTextEntry={false}
                  style={
                    Array [
                      Object {
                        "borderColor": "rgba(0,0,0,0)",
                        "borderRadius": 8,
                        "borderWidth": 2,
                        "color": "#9A9A9A",
                        "flex": 1,
                        "fontSize": 16,
                        "padding": 10,
                        "width": "100%",
                      },
                      Object {},
                      Object {},
                      Object {
                        "height": 50,
                      },
                      Object {},
                    ]
                  }
                  underlineColorAndroid="transparent"
                  value=""
                />
                <View />
                <View />
                <View
                  style={
                    Object {
                      "height": 200,
                      "width": "100%",
                      "zIndex": 5,
                    }
                  }
                >
                  <View />
                </View>
              </ScrollView>
            </View>
          </View>
        </ScrollView>
        <View
          style={
            Object {
              "alignSelf": "center",
              "flexDirection": "row",
              "marginBottom": 15,
              "marginTop": 0,
              "width": 712.5,
            }
          }
        >
          <TouchableOpacity
            activeOpacity={0.7}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderColor": "#3490dc",
                  "borderRadius": 4,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginVertical": 5,
                },
                Object {},
                Object {
                  "flex": 1,
                  "margin": 10,
                },
                Object {},
              ]
            }
            testID="clickable"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#3490dc",
                    "fontSize": 20,
                    "fontWeight": "700",
                    "padding": 6,
                  },
                  Object {},
                  Object {},
                  Object {},
                ]
              }
            >
              Back
            </Text>
          </TouchableOpacity>
          <TouchableOpacity
            activeOpacity={1}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#3490dc",
                  "borderRadius": 4,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginVertical": 5,
                },
                Object {
                  "alignItems": "center",
                  "backgroundColor": "grey",
                  "borderRadius": 4,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginVertical": 5,
                },
                Object {
                  "flex": 1,
                  "margin": 10,
                },
                Object {},
              ]
            }
            testID="clickable"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "fontWeight": "700",
                    "padding": 6,
                  },
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "fontWeight": "700",
                    "padding": 6,
                  },
                  Object {},
                  Object {},
                ]
              }
            >
              Next
            </Text>
          </TouchableOpacity>
        </View>
      </View>
    </View>
  </TouchableOpacity>
</View>
`;
